teps/0023-implicit-mapping.md

```diff
@@ -0,0 +1,358 @@
+              type: string
+          steps:
+            - name: echo
+              image: ubuntu
+              script: |
+                #!/usr/bin/env bash
+                echo "$(params.MESSAGE)"
+        params:
+          - name: MESSAGE
+            value: $(params.MESSAGE)
+  params:
+    - name: MESSAGE
+      value: "Good Morning!"
+```
+
+[Extra params are safe to include in Tasks](https://github.com/tektoncd/pipeline/blob/0593c7bef395505fae8b2e611cf4632e2e980e1d/examples/v1beta1/pipelineruns/pipelinerun-with-extra-params.yaml)
+since they don't modify behavior unless they are actually used - because of this
+there is little risk in passing through all parent param values to embedded
+specs.
+
+To avoid issues of additional webhook latency or reliability, we will not
+support remote TaskRefs or PipelineRefs.
+
+### Caveats
+
+#### Naming conflicts
+
+In cases of parameter naming conflicts, the innermost definition should win.
+e.g. for the following config:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  name: pipelinerun-with-taskspec-to-echo-message
+spec:
+  pipelineSpec:
+    tasks:
+      - name: echo-message
+        taskSpec:
+          params:
+            - name: MESSAGE
+              type: string
+          steps:
+            - name: echo
+              image: ubuntu
+              script: |
+                #!/usr/bin/env bash
+                echo "$(params.MESSAGE)"
+  params:
+    - name: MESSAGE
+      value: ["Good Morning!"]
+```
+
+The innermost Task definition should take precedence and make the config
+**invalid**, since the PipelineRun array param should not be able to override
+the param definition of the task.
+
+#### Explicit Definitions
+
+We should only define implicit params if no params section is defined. This
+ensures that this behavior will remain backwards compatible for existing
+defintions. e.g.
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  name: pipelinerun-with-taskspec-to-echo-message
+spec:
+  pipelineSpec:
+    params:
+      - name: MESSAGE
+        type: string
+    tasks:
+      - name: echo-message
+        taskSpec:
+          steps:
+            - name: echo
+              image: ubuntu
+              script: |
+                #!/usr/bin/env bash
+                echo "$(params.MESSAGE)"
+  params:
+    - name: MESSAGE
+      value: "Good Morning!"
+    - name: UNUSED
+      value: "unused message"
+```
+
+The `UNUSED` param will _not_ be plumbed through to the underlying embedded Task
+(even though it's an extra parameter to the Pipeline).
+
+#### Common parameter names
+
+Certain parameter names may be relatively common and therefore carry a different
+meaning between different `Tasks` (e.g. `url`, `path`, etc). `Tasks` with the
+same parameter names that require different input could therefore inadvertently
+acquire the wrong value.
+
+We consider it the `Pipeline` author's responsibility to guard against this.
```

Comment:
We could recommend Task authors to "namespace" params. 
eg: instead of `path`, `url` we can recomend `<task-name>_path` and `<task-name>_url`.
We can even make this a check in catalgo CI. 😇 

I think we need to specify best practices (or style guide) authoring as a different TEP. 🧑‍💻 📖

Reply:
This would probably not work well with kustomize/generateName, where names might be appended to.

Reply:
true, so no TaskName prefix, but any meaningful prefix chosen by Task manifest author will fit.